Additional responses can be status groups or "default"

divums opened the pull request:
Catching exception when generating OpenAPI specification. Added status groups and "default" as possible key for the additional responses.

If merged, this issue fixes #428

dmontagu commented on fastapi/openapi/utils.py:
```diff
@@ -190,12 +190,32 @@ def get_openapi_path(
+                            int(additional_status_code)
+                        )
+                    except ValueError as ex:
+                        status_groups: Dict[str, str] = {
```
Probably better to put the status groups dict at the module level (near `validation_error_definition`)

Then this can be much more concise:
```
status_text: Optional[str] = status_groups.get(str(additional_status_code).lower())
status_text = status_text or http.client.responses.get(int(additional_status_code))
```

divums replied:
Indeed it reads more clearly than the try-except block

dmontagu commented on Pipfile:
```diff
@@ -23,6 +23,9 @@ flake8 = "*"
 uvicorn = "*"
+bcrypt = "*"
+PyJWT = "*"
+passlib = "*"
```
Did you mean to include these?

divums replied:
when installing requirements via pip, those were missing and had to install manually in order to run the tests. i was waiting feedback whether that's correct or just my issue.

dmontagu replied:
They are only dev requirements, which is why I think they weren't included.

divums replied:
reverted this change. i suppose if nobody needed them so far, they were only issue because of the pip install i did

tiangolo replied:
Yeah, I'm not completely happy with Pipenv. It's OK like this for now.

I might switch the environment handling to pure `python -m venv`, Poetry or something. But that will be later.

divums committed:
Additional responses can be status groups or "default" (Fixes #428)

dmontagu commented on tests/test_additional_responses_bad.py:
```diff
@@ -0,0 +1,40 @@
+
+@app.get("/a", responses={"hello": {"description": "Not a valid additional response"}})
+async def a():
+    return "a"
```
```suggestion
    pass  # pragma: no cover
```
I think this will make the test coverage pass; this is how it's done elsewhere where the goal is not to test the actual endpoint.

divums replied:
i was in the middle of checking why the coverage dropped but haven't managed to reach any conclusion yet. thanks a lot, this comment saved a lot of time for me.

tiangolo committed:
:recycle: Refactor a bit status_code_ranges and use uppercase

...to conform with the spec: https://github.com/OAI/OpenAPI-Specification/blob/master/versions/3.0.2.md#patterned-fields-1

tiangolo committed:
:white_check_mark: Augment and update tests for response status codes with ranges

tiangolo commented:
Awesome, thanks @divums ! Good job! :rocket: :cake: 

I updated it a bit, to use uppercase instead of lowercase, to conform with the spec: https://github.com/OAI/OpenAPI-Specification/blob/master/versions/3.0.2.md#patterned-fields-1

And I augmented the tests to include some more of these cases.

And thanks a lot @dmontagu for the code review and help here! :clap: :muscle: :cake:

tiangolo commented:
I realized I introduced a bug making it `"DEFAULT"` instead of `"default"`, while making the status code ranges uppercase.

But I just fixed it in #489 :bug: 

I also added extra tests there to include `"default"` and other status codes at the same time and discovered another bug in the models. But it's fixed now. :sweat_smile: :rocket:
